synapse/storage/data_stores/main/events_bg_updates.py

```diff
@@ -503,3 +508,54 @@ def _event_fix_redactions_bytes_txn(txn):
         yield self._end_background_update("event_fix_redactions_bytes")
 
         return 1
+
+    @defer.inlineCallbacks
+    def _event_store_labels(self, progress, batch_size):
+        """Stores labels for events."""
+        last_event_id = progress.get("last_event_id", "")
+
+        def _event_store_labels_txn(txn):
+            txn.execute(
+                """
+                SELECT event_id, json FROM event_json
+                LEFT JOIN event_labels USING (event_id)
+                WHERE event_id > ? AND label IS NULL
+                LIMIT ?
+                """,
+                (last_event_id, batch_size)
+            )
+
+            rows = self.cursor_to_dict(txn)
+            if not rows:
+                return True
+
+            for row in rows:
+                event_id = row["event_id"]
+                event_json = json.loads(row["json"])
+
+                self._simple_insert_many_txn(
+                    txn=txn,
+                    table="event_labels",
+                    values=[
+                        {
+                            "event_id": event_id,
+                            "label": label,
+                        }
+                        for label in event_json["content"].get(LabelsField, [])
+                    ]
+                )
+
+            self._background_update_progress_txn(
+                txn, "event_store_labels", {"last_event_id": event_id}
+            )
+
+            return len(rows) == batch_size
```

Comment:
what? surely this will only make it run the first batch?

Reply:
Oh yes indeed. Now I'm only wondering why it didn't do that when I tested it on my labs homeserver...

Reply:
Weirdly enough, on my labs instance, it runs the whole thing with `==` and only once with `!=`. I've added a TODO for me to investigate tomorrow.

Reply:
That was an ordering issue, fixed in https://github.com/matrix-org/synapse/pull/6310/commits/f6356c4091f23c73f6944a475cb848ce047ab5b3